File: Java/DFS/SumOfPathNumbers.java
/*
*   @author
*   Aakash Verma
*
*   Output:
*   1008
*
*/

import java.util.*;

class Node {
    int data;
    Node left, right;
    public Node(int key) {
        data = key;
        left = right = null;
    }
}


class BinaryTreePathSum {

    static Node root;
    BinaryTreePathSum() {
        root = null;
    }

    private static int findPathSum(Node root, int pathSum) {
        if(root == null) return 0;

        pathSum = pathSum * 10 + root.data;

        if(root.left == null && root.right == null) {
            return pathSum;
        }

        return findPathSum(root.left, pathSum) + findPathSum(root.right, pathSum);
    }

    public static int findSumPathNumbers(Node root) {
        return findPathSum(root, 0);
    }

    public static void main (String[] args) {
        root = new Node(3);
        root.left = new Node(7);
        root.right = new Node(1);
        root.left.left = new Node(9);
        root.right.left = new Node(4);
        root.right.right = new Node(5);
        int ans = findSumPathNumbers(root);
        System.out.println(ans);
    }
}

File: Python/DFS/SumOfPathNumbers.py
# @author
# Aakash Verma

# Output:
# 1008

from collections import *

class Node:
    def __init__(self, data):
        self.data = data
        self.right = self.left = None

class BinaryTreePathSum:
    
    def __init__(self): 
        self.root = None

    def findPathSum(self, root, pathSum):
        if root is None:
            return 0

        pathSum = 10 * pathSum + root.data

        if root.left is None and root.right is None:
            return pathSum

        return self.findPathSum(root.left, pathSum) + self.findPathSum(root.right, pathSum)

    def findSumPathNumbers(self, root):
        return self.findPathSum(root, 0)
        
if __name__ == '__main__':
    tree = BinaryTreePathSum()
    tree.root = Node(3)
    tree.root.left = Node(7)
    tree.root.right = Node(1)
    tree.root.left.left = Node(9)
    tree.root.right.left = Node(4)
    tree.root.right.right = Node(5)
    ans = tree.findSumPathNumbers(tree.root)
    print(ans)


